File: minimax.py
from typing import Tuple
from classes import Piece, GameBoard
from possible_moves import get_all_possible_moves_for_team
from rules import make_move

def opposite(team:str, smash:bool):
    if smash: return team
    return "R" if team == "G" else "G"

def terminal(game_board: GameBoard) -> bool:
    """Endzustand ja oder nein?"""
    number_red_pieces = all(piece.team =="R" for row in game_board.game_board for piece in row) 
    number_green_pieces = all(piece.team =="G" for row in game_board.game_board for piece in row) 

    return number_red_pieces or number_green_pieces


def value_from(game_board: GameBoard):
    def value_from_piece(piece, positional_weight):
        piece_count_weight = 5
        kinged_piece_weight = 3
        return piece_count_weight + kinged_piece_weight * piece.king + piece.position[0] * positional_weight

    player_score = sum(
        value_from_piece(piece, positional_weight=0.5)
        for row in game_board.game_board
        for piece in row
        if piece.team == game_board.currPlayer
    )

    opponent_score = sum(
        value_from_piece(piece, positional_weight=0.5)
        for row in game_board.game_board
        for piece in row
        if piece.team != " " and piece.team != game_board.currPlayer
    )

    return opponent_score - player_score


def player(game_board: GameBoard):
    """Wer ist am Zug?
    return max or min player"""
    return game_board.currPlayer


def actions(game_board: GameBoard, team: str):
    """Welche Züge sind möglich?
    return list of possible moves"""
    pieces_with_all_possible_moves = get_all_possible_moves_for_team(game_board=game_board, team=team)
    actions = [(piece.position, move) for piece in pieces_with_all_possible_moves for move in piece.possible_positions]
    return actions
    


def result(game_board: GameBoard, action: Tuple[Tuple[int, int], Tuple[int, int]]):
    if not terminal(game_board=game_board):
        oldPosition, newPosition = action
        piece = game_board.game_board[oldPosition[0]][oldPosition[1]]
        return make_move(game_board=game_board, newPosition=newPosition, piece=piece)
    else: return game_board





def minimax(state:GameBoard, depth, player):

    best_score = [None,  -1000000] if player == "R" else [None, +1000000]

    if depth == 0 or terminal(state):
        score = value_from(state)
        return [None, score]

    for action in actions(state, player):
        result1 = result(state, action)
        value = minimax(result1, depth - 1, opposite(player, result1.currPlayer == player))

        if player == "R":
            if value[1] > best_score[1]:
                best_score = [action, value[1]]
        else:
            if value[1] < best_score[1]:
                best_score = [action, value[1]]

    return best_score

def minimax_for_gui(state:GameBoard, player):
    best = minimax(state, 4, player)
    return result(state, best[0])

File: Go/minimax/minimax.go
package minimax

import (
	"checkers/packages/gameboard"
	"checkers/packages/possible_moves"

	"github.com/samber/lo"
)

const (
	positionalWeight  = 0.5
	pieceCountWeight  = 5.0
	kingedPieceWeight = 3.0
)

type BestScore struct {
	Action gameboard.Action
	Score  float64
}

func opposite(team string, smash bool) string {
	if smash {
		return team
	}
	if team == "G" {
		return "R"
	}
	return "G"
}

func Terminal(gameBoard gameboard.GameBoard) bool {
	pieces := lo.Flatten(gameBoard.GameBoard)
	allLeftPiecesRed := lo.CountValuesBy(pieces, func(piece gameboard.Piece) bool {
		return piece.Team == "R" || piece.Team == " "
	})
	allLeftPiecesGreen := lo.CountValuesBy(pieces, func(piece gameboard.Piece) bool {
		return piece.Team == "G" || piece.Team == " "
	})
	moves := possible_moves.GetAllPossibleMovesForTeam(gameBoard, gameBoard.CurrPlayer)
	return allLeftPiecesRed[false] == 0 || allLeftPiecesGreen[false] == 0 || len(moves) == 0
}

func evaluateHeuristicValue(gameBoard gameboard.GameBoard) float64 {
	evalPlayer := func(player string) float64 {
		valueFromPiece := func(piece gameboard.Piece) float64 {
			king := 0.0
			if piece.King {
				king = 1.0
			}
			return pieceCountWeight + kingedPieceWeight*king + float64(piece.Position.Row)*positionalWeight
		}

		filteredPieces := lo.Filter(lo.Flatten(gameBoard.GameBoard), func(piece gameboard.Piece, _ int) bool {
			return player == piece.Team
		})

		return lo.Reduce(filteredPieces, func(agg float64, item gameboard.Piece, _ int) float64 {
			return agg + valueFromPiece(item)
		}, 0)
	}

	playerScore := evalPlayer(gameBoard.CurrPlayer)
	opponentScore := evalPlayer(opposite(gameBoard.CurrPlayer, false))

	return opponentScore - playerScore
}

func possibleActions(gameBoard gameboard.GameBoard, team string) []gameboard.Action {
	possibleMovePieces := possible_moves.GetAllPossibleMovesForTeam(gameBoard, team)
	return possible_moves.GetActionsFromPossibleMoves(gameBoard, possibleMovePieces)
}

func performAction(gameBoard gameboard.GameBoard, action gameboard.Action) gameboard.GameBoard {
	if !Terminal(gameBoard) {
		return possible_moves.MakeMove(gameBoard, action)
	}
	return gameBoard
}

func Minimax(state gameboard.GameBoard, depth int, player string) BestScore {
	bestScore := BestScore{Action: gameboard.Action{}, Score: -1000000}
	if player == "G" {
		bestScore.Score = 1000000
	}

	if depth == 0 || Terminal(state) {
		score := evaluateHeuristicValue(state)
		return BestScore{Action: gameboard.Action{}, Score: score}
	}

	for _, action := range possibleActions(state, player) {
		result := performAction(state, action)
		value := Minimax(result, depth-1, opposite(player, result.CurrPlayer == player))

		if player == "R" && value.Score > bestScore.Score {
			bestScore = BestScore{Action: action, Score: value.Score}
		} else if player == "G" && value.Score < bestScore.Score {
			bestScore = BestScore{Action: action, Score: value.Score}
		}
	}
	return bestScore
}
